forc-pkg/src/pkg.rs

```diff
@@ -450,19 +461,35 @@ fn validate_deps(
     graph: &Graph,
     node: NodeIx,
     node_manifest: &PackageManifestFile,
+    contract_dependencies: &mut HashSet<String>,
```

Comment:
Rather than passing in a `HashSet` of `contract_dependencies` here to determine whether or not dependencies are contract dependencies throughout the dependency validation process, we might want to describe the dependency *kind* in the `Edge` type.

Currently, the graph's `Edge` type is just the dependency name, however we could update it to also contain the dependency kind like so:

```rust
struct Edge {
    /// The name of the dependency as declared under `[dependencies]` or `[contract-dependencies]`.
    /// This may differ from the package name as declared under the dependency package's manifest.
    name: String,
    kind: DepKind,
}

enum DepKind {
    /// The dependency is a library and declared under `[dependencies]`.
    Library,
    /// The dependency is a contract and declared under `[contract-dependencies]`.
    Contract,
}
```

This should help to simplify a lot of the implementation involved here, as rather than having to repeatedly check if a name exists within a `contract_dependencies` set, we can instead simply check the edge's dependency kind when we need to distinguish behaviour.

Reply:
Thanks for the suggestion I think it is much more clearer right now!